ui/src/components/Status.vue

```diff
@@ -0,0 +1,221 @@
+      })
+      .catch((err) => {
+        console.error(err);
+        this.isError = true;
+      });
+    },
+    updateOverallState(state, progress){
+      // If the overallState is undefined, then it's the first load
+      if (!this.overallState) {
+        this.overallState = state;
+        this.$emit('updateOverallState', {state, progress: this.progress, isInit: true});
+      }
+
+      // When the state changed from non-X to X, we say the X action just
+      // finished, and update the parent.
+      else if (this.overallState !== 'OFF' && state === 'OFF') {
+        this.$emit('updateOverallState', {state});
+        this.log('Turned off successfully');
+        this.overallState = state;
+        this.setTransientTimer(false);
+      }
+      else if (this.overallState !== 'ON' && state === 'ON') {
+        this.$emit('updateOverallState', {state});
+        this.log('Turned on successfully');
+        this.overallState = state;
+        this.setTransientTimer(false);
+      }
+      // We keep send if the state is TRANSIENT, unless it's TIMEOUT
+      // TIMEOUT is a state for frontend only, it's created in setTransientTimer
+      else if (this.overallState !== 'TIMEOUT' && state === 'TRANSIENT') {
+        this.$emit('updateOverallState', {state, progress});
+        this.overallState = state;
+        this.setTransientTimer(true);
+      }
+      else if (this.overallState === state) {/* Do nothing */}
+      else if (this.overallState === 'TIMEOUT' && state === 'TRANSIENT') {/* Do nothing */}
+      else {
+        this.isError = true;
+        this.$emit('updateOverallState', {state});
+        this.log('Abnormal state' + state);
+        this.overallState = state;
+      }
+    },
+    countProgress(machines) {
+      let total = 0;
+      let finished = 0;
+      machines.forEach((machine) => {
+        machine.components.forEach((component) => {
+          if (component.status) total ++;
+          if (component.status === 'ON') finished ++;
+        })
+      });
+      return Math.round(finished / total * 100) + '%';
+    },
+    setTransientTimer(isStarting) {
+      // This is a function for controlling a timer, which tracks if the
+      // TRANSIENT state stay for longer than 1 min.
+      
+      // isStarting == false means the system is ON or OFF now. Stop the timeout timer
+      if (!isStarting) {
+        clearTimeout(this.transientTimer);
+        this.transientTimer = undefined;
+      }
+      // If isStarting (i.e. the state if TRANSIENT) and we don't have a timer,
+      // then we start a timer.
+      else if (!this.transientTimer) {
+        this.transientTimer = setTimeout(() => {
+          this.overallState = 'TIMEOUT'
+          this.$emit('updateOverallState', {state: 'TIMEOUT', progress: this.progress});
+        }, TIMEOUT);
+      }
+    },
+    updateEachDevice(machines) {
+      machines.forEach(machine => {
+        switch (machine.name) {
+          case 'SkyView - Left':
```

Comment:
what can we do to not hard-code machine names? It won't work if a deployment has different settings.

Reply:
Not easily. Currently all the machines are just flat in the config response. The only way I can tell the different between a skyview and other machines, and tell which direction it is, are the name.

It can be achieved if we change the config structure.